### PR TITLE
Test FakeRedis against real Redis

### PR DESCRIPTION
**Pull request by jerith**
There are a number of differences between the behaviours of FakeRedis and a real Redis client+server combination. We really should run the same set of tests for both and fix the differences.

### CODE REVIEW

**Comment by jerith** [comment]
Whoops, I left some tests commented. I'll fix that shortly.

File: vumi/persist/tests/test_fake_redis.py [hunk 1]
```diff
@@ -1,491 +1,747 @@
...
+
+    def wait(self, redis, delay):
+        redis._fake_redis.clock.advance(delay)
+        d = Deferred()
+        reactor.callLater(delay, d.callback, None)
+        return d
+
+
+# TestFakeRedisVerify.skip = "foo"
+# TestFakeRedisVerifyAsync.skip = "foo"
```
Comment (justinvdm) [inline vumi/persist/tests/test_fake_redis.py]: Should these be here?

Comment (jerith) [inline vumi/persist/tests/test_fake_redis.py]: Good catch, thanks. I'll remove them.

**Comment by justinvdm** [comment]
Would it be worth running the vumi-go tests using the fake redis changes? Not sure if that needs to be done before this PR lands though.

**Comment by jerith** [comment]
> Would it be worth running the vumi-go tests using the fake redis changes? Not sure if that needs to be done before this PR lands though.

Good idea, I'll do that now. Thanks!

**Comment by jerith** [comment]
vumi-go is happy with the new fake Redis. (Most of the changes were response values on write operations, so we were pretty much ignoring them anyway.)

**Comment by justinvdm** [comment]
> vumi-go is happy with the new fake Redis. (Most of the changes were response values on write operations, so we were pretty much ignoring them anyway.)

Awesome, thanks for checking.

File: vumi/persist/tests/test_fake_redis.py [hunk 2]
```diff
@@ -1,491 +1,750 @@
...
+    Each test in here should explain why verification isn't possible.
+    """
 
     @inlineCallbacks
     def test_scan_simple(self):
+        """
+        Scanning returns keys in an order that depends on arbitrary state in
+        the Redis server, so we can't fake it in a way that's identical to real
+        Redis.
+        """
```
Comment (justinvdm) [inline vumi/persist/tests/test_fake_redis.py]: I would've thought testing this against both fake and real redis is possible if we just sort the returned keys before asserting on them (along with a docstring explaining why sorting is necessary in the test). Any reason why that wouldn't work that I'm missing?

Comment (justinvdm) [inline vumi/persist/tests/test_fake_redis.py]: Oh, right, we can't assume that a set of keys returned by a scan call will always be the same set of keys. For some reason I assumed that would be the case, nevermind.

File: vumi/persist/tests/test_fake_redis.py [hunk 3]
```diff
@@ -1,491 +1,750 @@
...
+                errors.append(e)
+            else:
+                results.append(result)
+
+        # First, handle errors.
+        if errors:
+            if results:
+                self._test_case.fail(
+                    "One redis returned %r and the other raised %r" % (
+                        results[0], errors[0]))
```
Comment (justinvdm) [inline vumi/persist/tests/test_fake_redis.py]: It would be helpful to know which one returned something and which one raised something, might save us some confusion.

File: vumi/persist/tests/test_fake_redis.py [hunk 4]
```diff
@@ -1,491 +1,750 @@
...
+    def get_redis(self, **kwargs):
+        from vumi.persist.txredis_manager import TxRedisManager
+        # Fake redis
+        fake_redis = FakeRedis(async=True, **kwargs)
+        self.add_cleanup(fake_redis.teardown)
+        # Real redis
+        config = {
+            'FAKE_REDIS': 'yes',
+            'key_prefix': 'redistest',
+        }
```
Comment (justinvdm) [inline vumi/persist/tests/test_fake_redis.py]: If I'm understanding this correctly setting `FAKE_REDIS` to `'yes'` would cause us to create a fake redis instance, even though this config is for real redis. Any reason why we do that here?

File: vumi/persist/tests/test_fake_redis.py [hunk 5]
```diff
@@ -1,491 +1,750 @@
...
+                raise errors[0]
+
+        # Now handle results.
+        self._test_case.assertEqual(
+            results[0], results[1],
+            "Fake redis (a) and real redis (b) responses different:")
+        returnValue(results[0])
+
+    def __getattr__(self, name):
+        return lambda *args, **kw: self._perform_operation(name, *args, **kw)
```
Comment (justinvdm) [inline vumi/persist/tests/test_fake_redis.py]: I think we need tests for `FakeRedisWrapper`

Comment (justinvdm) [inline vumi/persist/tests/test_fake_redis.py]: err, s/`FakeRedisWrapper`/`RedisPairWrapper`

File: vumi/persist/txredis_manager.py [hunk 6]
```diff
@@ -154,6 +174,15 @@ def ttl(self, key):
         d.addCallback(lambda r: (None if r < 0 else r))
         return d
 
+    def persist(self, key):
+        self._send('PERSIST', key)
+        return self.getResponse()
+
+    def type(self, key):
+        d = self.get_type(key)
+        d.addCallback(lambda r: r if r is not None else 'none')
```
Comment (justinvdm) [inline vumi/persist/txredis_manager.py]: Any reason why we turn `None`s into `'none'`s? To ensure all results are string?

File: vumi/persist/txredis_manager.py [hunk 7]
```diff
@@ -154,6 +174,15 @@ def ttl(self, key):
         d.addCallback(lambda r: (None if r < 0 else r))
         return d
 
+    def persist(self, key):
+        self._send('PERSIST', key)
+        return self.getResponse()
```
Comment (justinvdm) [inline vumi/persist/txredis_manager.py]: Does txredis not implement this? If not, could we add a docstring mentioning that like we do above?

File: vumi/persist/txredis_manager.py (hunk 6 above)
Comment (justinvdm) [inline vumi/persist/txredis_manager.py]: Ah, is `'none'` maybe what the sync client returns?

**Comment by justinvdm** [comment]
I might just be not finding it in the diff, I don't think we have tests for `set` (including making sure that we turn `OK`s into `True`s). Could we add test(s) for it?

**Comment by justinvdm** [comment]
Looks good, left some comments.

File: vumi/persist/tests/test_fake_redis.py (hunk 2 above)
Comment (jerith) [inline vumi/persist/tests/test_fake_redis.py]: Yeah. I tried to rewrite the tests in a way that would let us walk the scan over the whole set of keys and check that we got the same set of results at the end, but that ended up being more effort than it's really worth.

File: vumi/persist/tests/test_fake_redis.py (hunk 3 above)
Comment (jerith) [inline vumi/persist/tests/test_fake_redis.py]: I thought about that, but my vague ideas made the code rather messier. I'll see if I can come up with a better mechanism.

File: vumi/persist/tests/test_fake_redis.py (hunk 4 above)
Comment (jerith) [inline vumi/persist/tests/test_fake_redis.py]: If the `VUMITEST_REDIS_DB` environment variable is set, this is overridden. At the moment, setting `FAKE_REDIS` doesn't change any behaviour, but that may change in the future. (For example, it might be useful to ensure that all Redis managers created in tests have it set, although I can't think of a practical way to do that at the moment.)

File: vumi/persist/tests/test_fake_redis.py (hunk 5 above)
Comment (jerith) [inline vumi/persist/tests/test_fake_redis.py]: I thought about it, but couldn't come up with a good way to test it that didn't involve test code of a similar level of complexity, which would then have to be tested itself, etc. I'll see if I can find a better way.

File: vumi/persist/txredis_manager.py (hunk 6 above)
Comment (jerith) [inline vumi/persist/txredis_manager.py]: Yes. There's code in txredis that turns `'none'` into `None`, but leaves all the other return values of the `TYPE` command as strings. I'll add a comment.

File: vumi/persist/txredis_manager.py (hunk 7 above)
Comment (jerith) [inline vumi/persist/txredis_manager.py]: Correct. I'll do that, thanks.

**Comment by jerith** [comment]
> I might just be not finding it in the diff, I don't think we have tests for `set` (including making sure that we turn `OK`s into `True`s). Could we add test(s) for it?

I'd prefer to leave that for another branch -- the purpose of this one is to build the verification infrastructure. I only added new tests for things that changed in ways that required modifications to other tests, like `lpush` and `rpush`. There are probably a bunch of other methods that aren't being tested explicitly (or perhaps even at all) and I don't want this branch to get too big.

File: vumi/persist/tests/test_fake_redis.py (hunk 5 above)
Comment (justinvdm) [inline vumi/persist/tests/test_fake_redis.py]: Ok cool, thanks. Happy for it to be a separate PR.

File: vumi/persist/txredis_manager.py (hunk 6 above)
Comment (justinvdm) [inline vumi/persist/txredis_manager.py]: +1, thanks

**Comment by justinvdm** [comment]
> I'd prefer to leave that for another branch -- the purpose of this one is to build the verification infrastructure. I only added new tests for things that changed in ways that required modifications to other tests, like lpush and rpush. There are probably a bunch of other methods that aren't being tested explicitly (or perhaps even at all) and I don't want this branch to get too big.

Happy for that to happen in a separate PR

File: vumi/persist/fake_redis.py [hunk 8]
```diff
@@ -285,9 +293,16 @@ def hvals(self, key):
...
+            value = self._data.get(key, {}).get(field, "0")
+        except AttributeError:
+            raise ResponseError("WRONGTYPE Operation against a key holding"
+                                " the wrong kind of value")
         # the int(str(..)) coerces amount to an int but rejects floats
-        value = int(value) + int(str(amount))
+        try:
+            value = int(value) + int(str(amount))
+        except (TypeError, ValueError):
+            raise ResponseError("value is not an integer or out of range")
```
Comment (justinvdm) [inline vumi/persist/fake_redis.py]: Do we have tests for these cases:
  - using `"0"` as the default field
  - `hincrby` against a key holding the wrong kind of value
  - `hincrby` with a non-integer or out of range value

If not, could we add some?

Comment (jerith) [inline vumi/persist/fake_redis.py]: Theses cases are all covered in `test_hincrby`.

Comment (justinvdm) [inline vumi/persist/fake_redis.py]: Thanks, missed those.

**Comment by justinvdm** [comment]
:+1: when travis is happy.

Should we make tickets for `RedisPairWrapper` and `set` tests?

**Comment by jerith** [comment]
I opened #926 for the test coverage improvements.
